### ISSUE TITLE
Constant GridServiceDeployer messages for 'daemon' services

### ISSUE THREAD

**Issue by colinmollenhour** [kontena/kontena#1862]
About every 15 seconds Kontena runs a slew of ServiceBalancerJob and GridServiceDeployer jobs. Here is an example of one for a single service, but it is happening for every service that uses the "daemon" strategy.

```
~ # grep -F 'firewall' GridServiceDeployer.log
I, [2017-02-16T21:04:58.795902 #17]  INFO -- ServiceBalancerJob: rebalancing service: primary/web/firewall
I, [2017-02-16T21:04:58.870030 #17]  INFO -- GridServiceDeployer: starting to deploy primary/web/firewall
I, [2017-02-16T21:04:59.828303 #17]  INFO -- GridServiceDeployer: deploying service instance primary/web/firewall-1 to node test3.local.lan
I, [2017-02-16T21:04:59.829489 #17]  INFO -- GridServiceDeployer: throttling service instance primary/web/firewall deploy because of min_health limit (1 instances in-progress)
I, [2017-02-16T21:05:02.097529 #17]  INFO -- GridServiceDeployer: deploying service instance primary/web/firewall-2 to node test3.local.lan-3
I, [2017-02-16T21:05:02.098794 #17]  INFO -- GridServiceDeployer: throttling service instance primary/web/firewall deploy because of min_health limit (1 instances in-progress)
I, [2017-02-16T21:05:02.762597 #17]  INFO -- GridServiceDeployer: service primary/web/firewall has been deployed
```

```
~ # grep -F 'ServiceBalancerJob' GridServiceDeployer.log
I, [2017-02-16T21:04:58.580197 #17]  INFO -- ServiceBalancerJob: rebalancing service: primary/redis-cache/lb
I, [2017-02-16T21:04:58.633333 #17]  INFO -- ServiceBalancerJob: rebalancing service: primary/redis-session/lb
I, [2017-02-16T21:04:58.703402 #17]  INFO -- ServiceBalancerJob: rebalancing service: primary/web/minify
I, [2017-02-16T21:04:58.795902 #17]  INFO -- ServiceBalancerJob: rebalancing service: primary/web/firewall
I, [2017-02-16T21:04:58.840695 #17]  INFO -- ServiceBalancerJob: rebalancing service: primary/web/php-fpm
I, [2017-02-16T21:04:58.941562 #17]  INFO -- ServiceBalancerJob: rebalancing service: primary/web/nginx-frontend
I, [2017-02-16T21:04:59.034143 #17]  INFO -- ServiceBalancerJob: rebalancing service: primary/web/nginx-admin
I, [2017-02-16T21:04:59.433670 #17]  INFO -- ServiceBalancerJob: rebalancing service: primary/galera/lb
I, [2017-02-16T21:05:18.580930 #17]  INFO -- ServiceBalancerJob: rebalancing service: primary/redis-cache/lb
```

The one thing these all have in common is they are "daemon" strategy. The "ha" strategy services do not seem to have the same issue.

Here is an example of one of the services:

```
primary/redis-session/lb:
  created: 2017-02-08T19:24:01.213Z
  updated: 2017-02-08T19:24:01.703Z
  stack: primary/redis-session
  state: deploying
  image: quay.io/***REDACTED***redis-session-lb-latest
  revision: 2
  stack_revision: 1
  stateful: no
  scaling: 1
  strategy: daemon
  deploy_opts:
    min_health: 0.8
  dns: lb.redis-session.primary.kontena.local
  affinity: 
    - label==app
  env: 
    - SERVICE_HOSTNAME=server
    - UPDATE_FREQUENCY=20
  net: bridge
  volumes:
    - /var/run/magento:/run/magento
  links: 
    - server
  instances:
    redis-session.lb-1:
      rev: 2017-02-16 21:19:01 UTC
      service_rev: 2
      node: test2.local.lan
      dns: lb-1.redis-session.primary.kontena.local
      ip: 10.81.128.21
      public ip: 147.***REDACTED***.149
      status: running
      exit code: 0
    redis-session.lb-2:
      rev: 2017-02-16 21:19:01 UTC
      service_rev: 2
      node: test3.local.lan
      dns: lb-2.redis-session.primary.kontena.local
      ip: 10.81.128.99
      public ip: 147.***REDACTED***.11
      status: running
      exit code: 0
    redis-session.lb-3:
      rev: 2017-02-16 21:18:41 UTC
      service_rev: 2
      node: test4.local.lan
      dns: lb-3.redis-session.primary.kontena.local
      ip: 10.81.128.42
      public ip: 147.***REDACTED***.173
      status: running
      exit code: 0
```

Note, at this instant the "state" was "deploying", but in the 15 seconds or so between the GridServiceDeployer logs it is usually "running".

In this service example there are three instances running and three nodes with label "app" so there shouldn't need to be anything deployed but the message "throttling service instance primary/web/firewall deploy because of min_health limit (1 instances in-progress)" seems to indicate Kontena thinks there is something to be done still?

**Comment by jakolehm** [comment]
Does this only happen when you combine affinity & daemon strategy?

**Comment by colinmollenhour** [comment]
I've never tried it without affinity because in my grid I have services that should only be run on certain nodes. I use daemon because some containers communicate through sockets and must always be on the same nodes so scaling doesn't really make sense in these cases and also the nodes are purpose built for specific services.

Let me know if there is more information I can get you, or if you like I can try removing the affinity on a service since this is just a testing cluster.

**Comment by matti** [comment]
Same here on three different grids. Repro: deploy stack matti/lb-ingress (daemon+affinity)

**Comment by colinmollenhour** [comment]
This issue makes it really hard to debug other issues because the logs are so congested.. Services are constantly going in and out of "deploying" and "running" states, etc.. Recently one of my containers using the daemon+affinity just disappeared and is not coming back...
From app1:
```
root@app1:~# docker logs kontena-agent | grep galera | tail
I, [2017-03-01T19:00:38.966444 #1]  INFO -- Kontena::ServicePods::Creator: service is up-to-date: galera.lb-1
I, [2017-03-01T19:00:55.933136 #1]  INFO -- Kontena::ServicePods::Creator: creating service: galera.lb-1
I, [2017-03-01T19:00:55.933741 #1]  INFO -- Kontena::Workers::ImagePullWorker: pulling image with credentials: quay.io/bss-llc/lg-prod-app:galera-lb-latest
I, [2017-03-01T19:00:56.898331 #1]  INFO -- Kontena::Workers::ImagePullWorker: pulled image: quay.io/bss-llc/lg-prod-app:galera-lb-latest
I, [2017-03-01T19:00:56.963961 #1]  INFO -- Kontena::ServicePods::Creator: service is up-to-date: galera.lb-1
I, [2017-03-01T19:01:25.094605 #1]  INFO -- Kontena::ServicePods::Creator: creating service: galera.lb-1
I, [2017-03-01T19:01:25.095117 #1]  INFO -- Kontena::Workers::ImagePullWorker: pulling image with credentials: quay.io/bss-llc/lg-prod-app:galera-lb-latest
I, [2017-03-01T19:01:26.213002 #1]  INFO -- Kontena::Workers::ImagePullWorker: pulled image: quay.io/bss-llc/lg-prod-app:galera-lb-latest
I, [2017-03-01T19:01:26.263270 #1]  INFO -- Kontena::ServicePods::Creator: service is up-to-date: galera.lb-1
I, [2017-03-01T19:01:56.927260 #1]  INFO -- Kontena::ServicePods::Terminator: terminating service: /galera.lb-1
```

From master:
```
I, [2017-03-01T19:01:14.811380 #10]  INFO -- ServiceBalancerJob: rebalancing service: primary/galera/lb
I, [2017-03-01T19:01:23.201944 #10]  INFO -- GridServiceDeployer: starting to deploy primary/galera/lb
I, [2017-03-01T19:01:24.985062 #10]  INFO -- GridServiceDeployer: deploying service instance primary/galera/lb-1 to node app1
I, [2017-03-01T19:01:24.986554 #10]  INFO -- GridServiceDeployer: throttling service instance primary/galera/lb deploy because of min
_health limit (1 instances in-progress)
I, [2017-03-01T19:01:31.542337 #10]  INFO -- GridServiceDeployer: deploying service instance primary/galera/lb-2 to node admin1
I, [2017-03-01T19:01:31.543198 #10]  INFO -- GridServiceDeployer: throttling service instance primary/galera/lb deploy because of min
_health limit (1 instances in-progress)
I, [2017-03-01T19:01:33.137224 #10]  INFO -- GridServiceDeployer: deploying service instance primary/galera/lb-3 to node app3
I, [2017-03-01T19:01:33.141580 #10]  INFO -- GridServiceDeployer: throttling service instance primary/galera/lb deploy because of min
_health limit (1 instances in-progress)
I, [2017-03-01T19:01:35.044729 #10]  INFO -- GridServiceDeployer: service primary/galera/lb has been deployed
I, [2017-03-01T19:01:35.048961 #10]  INFO -- GridServiceDeployer: removing service instance primary/app2/galera.lb-1
I, [2017-03-01T19:01:54.819608 #10]  INFO -- ServiceBalancerJob: rebalancing service: primary/galera/lb
I, [2017-03-01T19:01:54.975454 #6]  INFO -- GridServiceDeployer: starting to deploy primary/galera/lb
I, [2017-03-01T19:01:56.862119 #6]  INFO -- GridServiceDeployer: deploying service instance primary/galera/lb-1 to node app2
I, [2017-03-01T19:01:56.863967 #6]  INFO -- GridServiceDeployer: throttling service instance primary/galera/lb deploy because of min_
health limit (1 instances in-progress)
E, [2017-03-01T19:06:57.141588 #6] ERROR -- GridServiceInstanceDeployer: failed to deploy service instance primary/galera/lb-1 to nod
e app2
E, [2017-03-01T19:06:57.142786 #6] ERROR -- GridServiceDeployer: halting deploy of primary/galera/lb, one or more instances failed
```

I don't really know what's going on here but everything was fine until the GridServiceDeployer tried to re-deploy an image that as far as I can tell didn't need to be redeployed. It was not able to fix itself but I was able to fix it by restarting the service `kontena service restart galera/lb`.

Seems like this is a pretty significant bug since containers need to be stable and not being constantly redeployed..

**Comment by SpComb** [comment]
This is  #1895 

I agree that this is spammy in the logs and causes undue thrashing on the server, in addition to constant Docker image pulls: #1731

The mitigating factor for why we didn't rush to fix this in the 1.1.x releases is that it should not affect the actual service containers themselves, since the agent should always consider them as being up to date:

```
Kontena::ServicePods::Creator: service is up-to-date: ....
```

However, your disappearing `galera.lb-1` service on the `app2` node is a bit odd. The `GridServiceDeployer` decided to remove it, but then the following `ServiceBalancerJob` immediately tried to re-deploy it:

```
I, [2017-03-01T19:01:14.811380 #10]  INFO -- ServiceBalancerJob: rebalancing service: primary/galera/lb
I, [2017-03-01T19:01:35.048961 #10]  INFO -- GridServiceDeployer: removing service instance primary/app2/galera.lb-1
I, [2017-03-01T19:01:54.819608 #10]  INFO -- ServiceBalancerJob: rebalancing service: primary/galera/lb
I, [2017-03-01T19:01:56.862119 #6]  INFO -- GridServiceDeployer: deploying service instance primary/galera/lb-1 to node app2
E, [2017-03-01T19:06:57.141588 #6] ERROR -- GridServiceInstanceDeployer: failed to deploy service instance primary/galera/lb-1 to node app2
```

Looking at the timestamps, the following `ServiceBalancerJob` -> `GridServiceDeployer` -> `GridServiceInstanceDeployer` happens at the same time as the agent `Kontena::ServicePods::Terminator` run from the previous deployment happens, and this confusion leads to the `GridServiceInstanceDeployer` timeout 5 minutes later.

This sounds like some separate issue with the actual `GridScheduler`. Were you possibly modifying the `kontena node label ...` settings when this happaned?

**Comment by SpComb** [comment]
Can you give the complete `kontena-agent` logs from the `app2` node around the time this happaned (`2017-03-01T19:01:56`)?

**Comment by colinmollenhour** [comment]
Sorry, I don't have those specific logs anymore but I think the issue was probably the wait_for_port option and the failure of that check to pass due to some sort of networking issue as I reported on #1895 [comment](https://github.com/kontena/kontena/issues/1895#issuecomment-288517692). I think the constant redeploying just makes this issue much more likely to happen.

**Comment by SpComb** [comment]
Re https://github.com/kontena/kontena/issues/1895#issuecomment-288517692:

> On the service that is broken if I use "nc -z" against the bridge IP it works and if I use it against the weave IP it doesn't.

Can you show the output of `kontena service exec --instance=1 redis-cache/lb ip addr`, and then while you are attempting to connect to it, the output of `ip neigh | grep 10.81.128.60` on the host machine / agent.

This almost sounds like it could be https://github.com/weaveworks/weave/issues/2842. Are you running the `kontena/vpn` service? If you suspect that, then try running the [`weave-zombie-hunter.sh`](https://gist.github.com/SpComb/9cc2a7ed46ff708547d09854b78c197f#file-weave-zombie-hunter-sh) script on the host node where the VPN service is running on.

**Comment by colinmollenhour** [comment]
Ahh, it is not broken now since I restarted docker daemon to get it working again. I currently don't have the issue and I don't know how to reproduce it other than waiting for it to happen again... Here is the output anyway **just in case** you still want it, and if/when it does happen again I'll get the output before restarting docker next time:

```
$ kontena service exec --instance=1 redis-cache/lb ip addr
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN qlen 1
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host 
       valid_lft forever preferred_lft forever
2162: eth0@if2163: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1500 qdisc noqueue state UP 
    link/ether 02:42:ac:11:00:02 brd ff:ff:ff:ff:ff:ff
    inet 172.17.0.2/16 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::42:acff:fe11:2/64 scope link 
       valid_lft forever preferred_lft forever
2164: ethwe@if2165: <BROADCAST,MULTICAST,UP,LOWER_UP,M-DOWN> mtu 1410 qdisc noqueue state UP 
    link/ether ba:98:3e:32:51:94 brd ff:ff:ff:ff:ff:ff
    inet 10.81.128.29/16 scope global ethwe
       valid_lft forever preferred_lft forever
    inet6 fe80::b898:3eff:fe32:5194/64 scope link 
       valid_lft forever preferred_lft forever
```

And from the host running redis-cache.lb-1:

```
~ # nc -z 10.81.128.29 6379 && echo yes
yes
~ # ip neigh | grep 10.81.128.29
10.81.128.29 dev weave lladdr ba:98:3e:32:51:94 REACHABLE
~ # docker exec -it kontena-agent sh
/app # ip neigh | grep 10.81.128.29
10.81.128.29 dev weave lladdr ba:98:3e:32:51:94 ref 1 used 0/0/0 probes 1 REACHABLE
```

>  Are you running the kontena/vpn service?

I am not/was not using Kontena VPN and this grid does not have a VPN service installed.

**Comment by SpComb** [comment]
> Here is the output anyway just in case you still want it, and if/when it does happen again I'll get the output before restarting docker next time:

Yeah, the interesting part is that the `ethwe link/ether ba:98:3e:32:51:94 inet 10.81.128.29` addresses match up with the `ip neigh` `10.81.128.29 ba:98:3e:32:51:94`.

The thing to watch out for re https://github.com/weaveworks/weave/issues/2842 is if the ethernet addresses do not match. If that does happen, then the newest clues for https://github.com/weaveworks/weave/issues/2842#issuecomment-288791826 point at the `docker-proxy` being responsible.

**Comment by jnummelin** [comment]
#1873 fixed the original issue and will be released with 1.2.0. Zombie network issues are tracked on separate issues.
